autopilot/contractor.go

```diff
@@ -1620,6 +1606,32 @@ func endHeight(cfg api.AutopilotConfig, currentPeriod uint64) uint64 {
 	return currentPeriod + cfg.Contracts.Period + cfg.Contracts.RenewWindow
 }
 
+func randSelectHosts(candidates map[types.PublicKey]scoredHost, wanted int) (selected []hostdb.Host) {
+	// turn map into slices
+	scores := make([]float64, 0, len(candidates))
+	scored := make([]hostdb.Host, 0, len(candidates))
+	for _, h := range candidates {
+		scored = append(scored, h.host)
+		scores = append(scores, h.score)
+	}
+
+	// select hosts
+	for len(selected) < wanted && len(scored) > 0 {
+		i := randSelectByWeight(scores)
+		selected = append(selected, scored[i])
+
+		// remove selected host
+		scored[i], scored = scored[len(scored)-1], scored[:len(scored)-1]
+		scores[i], scores = scores[len(scores)-1], scores[:len(scores)-1]
+	}
+
+	// sort by score
+	sort.Slice(selected, func(i, j int) bool {
```

Comment:
This is ruining the weighted selection we have just done. A host with a lower score already has a lower chance of being in front of a host with higher score.